doc/rm/asm_coding_style.md

```diff
@@ -0,0 +1,286 @@
+
+Some assemblers allow `la` with an immediate expression instead of a symbol, allowing a form of symbol+offset.
+However, support for this behavior is patchy, and the semantics of PIC `la` with immediate are unclear (in PIC mode, `la` should perform a GOT lookup, not a `pc`-relative load).
+
+### Jumping into C
+
+***Jumping into a C function must be done either with a `call` instruction, or, if that function is marked `noreturn`, a `tail` instruction.***
+
+The RISC-V jump instructions take a "link register", which holds the return address (this should always be `zero` or `ra`), and a small `pc`-relative immediate.
+For jumping to a symbol, there are two user-controlled settings: "near" or "far", and "returnable" (i.e., a link register of `zero` or `ra`).
+The mnemonics for these are:
+- `j sym`, for a near non-returnable jump.
+- `jal sym`, for a near returnable jump.
+- `tail sym`, for a far non-returnable jump (i.e., a non-unwinding tail-call).
+- `call sym`, for a far returnable jump (i.e., function calls).
+
+Far jumps are implemented in the assembler by emiting `auipc` instructions as necessary (since the jump-and-link instruction takes only a small immediate).
+Jumps into C should always be treated as far jumps, and as such use the `call` instruction, unless the C function is marked `noreturn`, in which case `tail` can be used.
+
+Example:
+```S
+  call _syscall_start
+
+  tail _crt0
+```
+
+### Control and Status Register (CSR) Names
+
+***CSRs defined in the RISC-V spec must be refered to by those names (like `mstatus`), while custom non-standard ones must be encapsulated in a `#define`.***
+
+Naturally, if a pseudoinstruction exists to read that CSR (like `rdtime`) that one should be used, instead.
+
+`#define`s for CSRs should be prefixed with `CSR_<design>_`, where `<design>` is the name of the design the CSR corresponds to.
+
+Recognized CSR prefixes:
+- `CSR_IBEX_` - A CSR specific to the Ibex core.
+- `CSR_OT_` - A CSR specific to the OpenTitan chip, beyond the Ibex core.
+
+Example:
+```S
+  csrr t0, mstatus
+
+  #define CSR_OT_HMAC_ENABLED ...
+  csrw CSR_OT_HMAC_ENABLED, 0x1
+```
+
+### Load and Store From Pointer in Register
+
+***When loading and storing from a pointer in a register, prefer to use `n(reg)` shorthand.***
+
+In the case that a pointer is being read without an offset, prefer `0(reg)` over `(reg)`.
+
+```S
+  // Correct:
+  lw t3, 8(sp)
+  sb t3, 0(a0)
+  // Wrong:
+  lw t3, sp, 8
+  sb t3, a0
+```
+
+### Compressed Instruction Mnemonics
+
+***Do not use compressed instruction mnemonics.***
+
+While Ibex implements the RISC-V C extension, it is expected that the toolchain will automatically compress instructions where possible.
+This advice extends to the `c.unimp` all-zeroes illegal instruction.
+Toolchains like `LLVM` and `gas` are expected to recognize the `unimp` mnemonic to stand for two `c.unimp` instructions.
```

Comment:
Actually, `unimp` is compressed to a single `c.unimp`. These two lines around `unimp` perhaps aren't necessary in this doc?

Reply:
Done.